pkg/infrastructure/aws/clusterapi/aws.go

```diff
@@ -262,52 +271,61 @@ func removeSSHRule(ctx context.Context, cl k8sClient.Client, infraID string, aws
 	}
 	logrus.Debug("Updated AWSCluster to remove bootstrap SSH rule")
 
-	key := k8sClient.ObjectKey{
-		Name:      infraID,
-		Namespace: capiutils.Namespace,
+	var sgID string
+	if sg, ok := awsCluster.Status.Network.SecurityGroups[capa.SecurityGroupControlPlane]; ok && len(sg.ID) > 0 {
+		sgID = sg.ID
+	} else if ok {
+		return fmt.Errorf("control plane security group id is not populated in awscluster status")
+	} else {
+		keys := make([]capa.SecurityGroupRole, 0, len(awsCluster.Status.Network.SecurityGroups))
+		for sgr := range awsCluster.Status.Network.SecurityGroups {
+			keys = append(keys, sgr)
+		}
+		return fmt.Errorf("controlplane not found in cluster security groups: %v", keys)
 	}
+
 	timeout := 15 * time.Minute
-	untilTime := time.Now().Add(timeout)
-	warnTime := time.Now().Add(5 * time.Minute)
-	warned := false
+	startTime := time.Now()
+	untilTime := startTime.Add(timeout)
 	timezone, _ := untilTime.Zone()
-	logrus.Infof("Waiting up to %v (until %v %s) for bootstrap SSH rule to be destroyed...", timeout, untilTime.Format(time.Kitchen), timezone)
-	if err := wait.ExponentialBackoffWithContext(ctx, wait.Backoff{
-		Duration: time.Second * 10,
-		Factor:   float64(1.5),
-		Steps:    32,
-		Cap:      timeout,
-	}, func(ctx context.Context) (bool, error) {
-		c := &capa.AWSCluster{}
-		if err := cl.Get(ctx, key, c); err != nil {
-			return false, err
-		}
-		if time.Now().After(warnTime) && !warned {
-			logrus.Warn("Deleting bootstrap SSH rule is still progressing but taking longer than expected")
-			warned = true
-		}
-		if sg, ok := c.Status.Network.SecurityGroups[capa.SecurityGroupControlPlane]; ok {
-			for _, r := range sg.IngressRules {
-				if r.Description == awsmanifest.BootstrapSSHDescription {
-					logrus.Debugf("Still waiting for bootstrap SSH security rule %s to be deleted from %s...", r.Description, sg.ID)
-					return false, nil
+	logrus.Debugf("Waiting up to %v (until %v %s) for bootstrap SSH rule to be destroyed...", timeout, untilTime.Format(time.Kitchen), timezone)
+	if err := wait.PollUntilContextTimeout(ctx, 15*time.Second, timeout, true,
+		func(ctx context.Context) (bool, error) {
+			sgs, err := awsconfig.DescribeSecurityGroups(ctx, session, []string{sgID}, region)
+			if err != nil {
+				return false, fmt.Errorf("error getting security group: %w", err)
+			}
+
+			if len(sgs) != 1 {
+				ids := []string{}
+				for _, sg := range sgs {
+					ids = append(ids, *sg.GroupId)
 				}
+				return false, fmt.Errorf("expected exactly one security group with id %s, but got %v", sgID, ids)
```

Comment:
It appears that the above for loop exists just to be able to print a list of security groups associated with the 1 sgID. Is it not sufficient to return error with the number of security groups i.e len(sgs)?

Reply:
I think we do something similar to this already. I think this is helpful to see what security groups were actually attached and which was expected. But either way will definitely work.

Reply:
There is a 1 o 1 correlation between security group id and security group, so this should never happen. This is just a defensive safeguard, and we include the ids for debugging in case this somehow did happen.